src/java.base/share/classes/java/lang/runtime/StringTemplateImpl.java

```diff
@@ -0,0 +1,130 @@
+ * string template bootstrap method callsites generated by the compiler. Instances of this
+ * class are produced by {@link StringTemplateImplFactory}.
+ * <p>
+ * Values are stored by subclassing {@link Carriers.CarrierObject}. This allows specializations
+ * and sharing of value shapes without creating a new class for each shape.
+ * <p>
+ * {@link StringTemplate} fragments are shared via binding to the
+ * {@link java.lang.invoke.CallSite CallSite's} {@link MethodHandle}.
+ * <p>
+ * The {@link StringTemplateImpl} instance also carries
+ * specialized {@link MethodHandle MethodHandles} for producing the values list and interpolation.
+ * These {@link MethodHandle MethodHandles} are  also shared by binding to the
+ * {@link java.lang.invoke.CallSite CallSite}.
+ *
+ * @since 21
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+final class StringTemplateImpl extends Carriers.CarrierObject implements StringTemplate {
+    /**
+     * List of string fragments for the string template. This value of this list is shared by
+     * all instances created at the {@link java.lang.invoke.CallSite CallSite}.
+     */
+    private final List<String> fragments;
+
+    /**
+     * Specialized {@link MethodHandle} used to implement the {@link StringTemplate StringTemplate's}
+     * {@code values} method. This {@link MethodHandle} is shared by all instances created at the
+     * {@link java.lang.invoke.CallSite CallSite}.
+     */
+    private final MethodHandle valuesMH;
+
+    /**
+     * Specialized {@link MethodHandle} used to implement the {@link StringTemplate StringTemplate's}
+     * {@code interpolate} method. This {@link MethodHandle} is shared by all instances created at the
+     * {@link java.lang.invoke.CallSite CallSite}.
+     */
+    private final MethodHandle interpolateMH;
+
+    /**
+     * Constructor.
+     *
+     * @param primitiveCount  number of primitive slots required (bound at callsite)
+     * @param objectCount     number of object slots required (bound at callsite)
+     * @param fragments       list of string fragments (bound in (bound at callsite)
+     * @param valuesMH        {@link MethodHandle} to produce list of values (bound at callsite)
+     * @param interpolateMH   {@link MethodHandle} to produce interpolation (bound at callsite)
+     */
+    StringTemplateImpl(int primitiveCount, int objectCount,
+                       List<String> fragments, MethodHandle valuesMH, MethodHandle interpolateMH) {
+        super(primitiveCount, objectCount);
+        this.fragments = fragments;
+        this.valuesMH = valuesMH;
+        this.interpolateMH = interpolateMH;
+    }
+
+    @Override
+    public List<String> fragments() {
+        return fragments;
+    }
+
+    @Override
+    public List<Object> values() {
+        try {
+            return (List<Object>)valuesMH.invokeExact(this);
+        } catch (Throwable ex) {
```

Comment:
Errors likes OutOfMemoryError and runtime exception should be rethrown instead of being wrapped

Reply:
Changing